### PR TITLE
Test script

### PR DESCRIPTION
**Pull request by aaishwarymishra**
This pull request introduces support for test splits and evaluation in the training workflow, alongside a refactor of the transformer model's attention mechanism and some minor code and documentation improvements. The most significant changes are the addition of a `test` split to the configuration, corresponding updates to the trainer to support test evaluation and logging, and the replacement of PyTorch's built-in multi-head attention with a custom implementation.

**Test split and evaluation support:**

- Added a `test` split to `DataConfig` in `trainite/config/base.py`, allowing users to specify a test dataset in configuration. The README and project initialization logic were updated to reflect and support this change. [[1]](diffhunk://#diff-73352dfb97a0b9afb3ad346b22af2fdb5df6e72c73da6aa0df96d85c0a4c0cddR45) [[2]](diffhunk://#diff-ada0f955d81307b406b77a5ab5e7423974538138ae1ead8e4005cce3b61cd101L42-R49) [[3]](diffhunk://#diff-1e20c19b0f95a48591a0f35c025ef834b85ff6c4deaecb675a711f8499f1bf40L330-L337)
- Updated `PreTrainer` in `trainite/trainers/pretrainer.py` to build and use a test dataloader, attach test metrics, run test evaluation after training, and log results to TensorBoard. Added a `test` method to handle loading the best model and running evaluation. [[1]](diffhunk://#diff-8f40d97481e59dd82f8477965646f5d9517f80b514ebf0d45e17346dae39c1caR36) [[2]](diffhunk://#diff-8f40d97481e59dd82f8477965646f5d9517f80b514ebf0d45e17346dae39c1caR63-R68) [[3]](diffhunk://#diff-8f40d97481e59dd82f8477965646f5d9517f80b514ebf0d45e17346dae39c1caR95) [[4]](diffhunk://#diff-8f40d97481e59dd82f8477965646f5d9517f80b514ebf0d45e17346dae39c1caR172-R188) [[5]](diffhunk://#diff-8f40d97481e59dd82f8477965646f5d9517f80b514ebf0d45e17346dae39c1caR289-R295) [[6]](diffhunk://#diff-8f40d97481e59dd82f8477965646f5d9517f80b514ebf0d45e17346dae39c1caR324-R349) [[7]](diffhunk://#diff-8f40d97481e59dd82f8477965646f5d9517f80b514ebf0d45e17346dae39c1caR366-R368)

**Transformer model improvements:**

- Replaced the use of PyTorch's `nn.MultiheadAttention` with a custom `Attention` class in `trainite/models/transformer.py`, providing more direct control over masking and dropout. Updated the transformer block and model forward passes to use the new attention mechanism and handle padding masks correctly. [[1]](diffhunk://#diff-23286b30be7f6477ea835453fdb3dcc78f5fe95323349abe4f0c6ebf17fe04e8R27-R79) [[2]](diffhunk://#diff-23286b30be7f6477ea835453fdb3dcc78f5fe95323349abe4f0c6ebf17fe04e8L37-R90) [[3]](diffhunk://#diff-23286b30be7f6477ea835453fdb3dcc78f5fe95323349abe4f0c6ebf17fe04e8L49-R104) [[4]](diffhunk://#diff-23286b30be7f6477ea835453fdb3dcc78f5fe95323349abe4f0c6ebf17fe04e8R142-R150)

**Other improvements and fixes:**

- Removed the default `learning_rate` from `PreTrainerConfig` in `trainite/config/trainer.py` to defer learning rate specification elsewhere.
- Minor code cleanup and import reordering in `trainite/datasets/string_reverse.py`, and ensured special tokens are included in the character-to-id mapping. [[1]](diffhunk://#diff-e0dac8bceead0c2b475a6440a83d3cb3f966218558961b9832d2eeb75dbb4250L2-R3) [[2]](diffhunk://#diff-e0dac8bceead0c2b475a6440a83d3cb3f966218558961b9832d2eeb75dbb4250L30) [[3]](diffhunk://#diff-e0dac8bceead0c2b475a6440a83d3cb3f966218558961b9832d2eeb75dbb4250R39-R41)

### CODE REVIEW

File: trainite/trainers/pretrainer.py [hunk 1]
```diff
@@ -307,6 +326,32 @@ def _run_evaluations(self, engine: Engine) -> None:
             val_metrics["token_accuracy"],
         )
 
+    def test(self, test_loader: DataLoader | None = None) -> None:
+        loader = test_loader or self.test_loader
+        if loader is None:
+            logger.warning("No test loader provided. Skipping testing.")
```
Comment (TahaZahid05) [inline trainite/trainers/pretrainer.py]: use self.logger

File: trainite/trainers/pretrainer.py [hunk 2]
```diff
@@ -307,6 +326,32 @@ def _run_evaluations(self, engine: Engine) -> None:
...
+            return
+
+        # Load best model if available
+        checkpoint_handler = self.handlers.get("checkpoint_best")
+        if checkpoint_handler and checkpoint_handler.last_checkpoint:
+            checkpoint_path = checkpoint_handler.last_checkpoint
+
+            logger.info("Loading best model for testing from %s", checkpoint_path)
```
Comment (TahaZahid05) [inline trainite/trainers/pretrainer.py]: .

File: trainite/trainers/pretrainer.py [hunk 3]
```diff
@@ -307,6 +326,32 @@ def _run_evaluations(self, engine: Engine) -> None:
...
+            checkpoint = torch.load(
+                checkpoint_path, map_location=self.device, weights_only=True
+            )
+            self.model.load_state_dict(checkpoint["model"])
+
+        logger.info("Running testing...")
```
Comment (TahaZahid05) [inline trainite/trainers/pretrainer.py]: .

File: trainite/trainers/pretrainer.py [hunk 4]
```diff
@@ -307,6 +326,32 @@ def _run_evaluations(self, engine: Engine) -> None:
...
+        self.test_evaluator.run(loader)
+        metrics = self.test_evaluator.state.metrics
+        logger.info(
```
Comment (TahaZahid05) [inline trainite/trainers/pretrainer.py]: .

**Comment by TahaZahid05** [comment]
You may also have to update the output_transforms of test if we merge my PR first.

**Comment by aaishwarymishra** [comment]
my bad I think i forgot to take the latest commit

**Review by TahaZahid05 — APPROVED** [review]

**Review by TahaZahid05 — APPROVED** [review]
